codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll

```diff
@@ -0,0 +1,221 @@
+/**
+ * Provides classes for identifying and reasoning about Microsoft source code
+ * annotation language (SAL) macros.
+ * 
+ * This version of this file has been updated to include driverspecs.h as a valid SAL header.
+ * It will be removed if and when that change is upstreamed to the main CodeQL repo.
+ */
+
+import cpp
+
+/**
+ * A SAL macro defined in `sal.h` or a similar header file.
+ */
+class SALMacro extends Macro {
+  SALMacro() {
+    this.getFile().getBaseName() =
+      ["sal.h", "specstrings_strict.h", "specstrings.h", "w32p.h", "minwindef.h", "driverspecs.h"] and
+    (
+      // Dialect for Windows 8 and above
+      this.getName().matches("\\_%\\_")
+      or
+      // Dialect for Windows 7
+      this.getName().matches("\\_\\_%")
+    )
+  }
+}
+
+pragma[noinline]
+private predicate isTopLevelMacroAccess(MacroAccess ma) { not exists(ma.getParentInvocation()) }
+
+/**
+ * An invocation of a SAL macro (excluding invocations inside other macros).
+ */
+class SALAnnotation extends MacroInvocation {
+  SALAnnotation() {
+    this.getMacro() instanceof SALMacro and
+    isTopLevelMacroAccess(this)
+  }
+
+  /** Gets the `Declaration` annotated by `this`. */
+  Declaration getDeclaration() {
+    annotatesAt(this, result.getADeclarationEntry(), _, _) and
+    not result instanceof Type // exclude typedefs
+  }
+
+  /** Gets the `DeclarationEntry` annotated by `this`. */
+  DeclarationEntry getDeclarationEntry() {
+    annotatesAt(this, result, _, _) and
+    not result instanceof TypeDeclarationEntry // exclude typedefs
+  }
+}
+
+/**
+ * A SAL macro indicating that the return value of a function should always be
```

Comment:
I'd actually like to push back against this change for two reasons:

1. This file is taken from the main CodeQL repo - it's only duplicated here because we need to include "driverspecs.h" as a source of annotations.  So I would prefer not to re-write comments taken directly from the main repo.  
2. The classes are the CodeQL representations of the SAL macros.  Looking at other CodeQL core libraries, this style of comment (where a class is simply described as what it is in the code being analyzed, rather than as "getting" that thing) is common.  See for example: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/Union.qll

Reply:
If this file is coming directly from the CodeQL repo then why is it copied here and not referenced from the mirrored codeql repo located at Windows-Driver-Developer-Supplemental-Tools/codeql/codeql-queries @ e178626 ?

Reply:
Because it was edited to add driverspecs.h to the list and is not a 1:1 copy.  I tried originally just to extend this class in a file we own, but unfortunately CodeQL inheritance doesn't work that way, so for now we need to have our own copy with our edits.

Reply:
If it is a straight copy then I can understand not wanting to change it at all, but if you are making changes to it anyway then why not update it to be clearer? I doubt we will be RI-ing anyway.
I'm not going to block on this but I believe it is worth considering.

Reply:
It is always ideal to not have to copy files.  But sometimes it is a reasonable pragmatic choice to copy.    (Typically, that then comes and bites you about 5 years later :-) )

In the case where you copy a file I am personally a fan of "minimizing the diff".  So, make changes as clean as possible for people in future running "windiff orig patched".  It is also reasonable to write a comment providing this context in the file"/* This file copied from x\y\z and only following line(s) was edited */".  That line will show up yellow of course, but the intension will be immediately clear.

My 2 cents.